### PR TITLE
Const pad support to pad2d and slice

### PR DESCRIPTION
**Pull request by ZenginU**
This PR, refactors pad2d and slice, eventually slice and pad2d does the same thing both can shrink and pad with zeros but they do it with different formatted input. So I have merged them into one in pad2d and used it in slice.

As a result of this PR, pad2d now supports const padding

To sum up, slice and pad2d does the same thing eventually

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/tensor.py                         6               8
------------------------------------------------------------
lines added in the tinygrad folder: -2
```

**Comment by wozeparrot** [comment]
Where is `Tensor.slice` actually used? Is it possible just to get rid of it entirely if pad2d essentially does what it does?

**Comment by ZenginU** [comment]
Currently it is being used in _pool and in onnx I believe, but remember that it does the same thing with different input so it may require some input formatting to replace slice with pad2d entirely.

For example, Tensor.ones(5).slice(((0,6),)) == Tensor.ones(5).pad2d((0,1))

**Comment by ZenginU** [comment]
Want me to migrate all slice to pad2d in this PR or in a seperate PR? Maybe if this gets accepted I can do it in sep PR

File: tinygrad/tensor.py [hunk 1]
```diff
@@ -251,10 +251,7 @@ def pad(self, arg: Tuple[Tuple[int, int], ...], value:float=0) -> Tensor:
   # ***** movement hlops *****
 
   # NOTE: using slice is discouraged and things should migrate to pad and shrink
-  def slice(self, arg:Sequence[Optional[Tuple[int, int]]]) -> Tensor:
-    arg_ = tuple([a if a is not None else (0,s) for s,a in zip(self.shape, arg)])
-    padding = tuple([(max(0, -p[0]), max(0, p[1]-self.shape[i])) for i,p in enumerate(arg_)])
-    return self.pad(padding).shrink(tuple([(p[0] + padding[i][0], p[1] + padding[i][0]) for i,p in enumerate(arg_)]))
+  def slice(self, arg: Sequence[Optional[Tuple[int, int]]]) -> Tensor: return self.pad2d(flatten(((a[1] - s, -a[0]) if a is not None else (0, 0) for s, a in zip(self.shape, arg)))[::-1])
```
Comment (geohot) [inline tinygrad/tensor.py:254]: slice should not call pad2d. Can this be rewritten so pad2d calls slice?

Comment (ZenginU) [reply tinygrad/tensor.py]: Current implementation is like that so I reverted my commit and added const support to both slice and pad2d

File: tinygrad/tensor.py [hunk 2]
```diff
@@ -251,10 +251,10 @@ def pad(self, arg: Tuple[Tuple[int, int], ...], value:float=0) -> Tensor:
   # ***** movement hlops *****
 
   # NOTE: using slice is discouraged and things should migrate to pad and shrink
-  def slice(self, arg:Sequence[Optional[Tuple[int, int]]]) -> Tensor:
-    arg_ = tuple([a if a is not None else (0,s) for s,a in zip(self.shape, arg)])
-    padding = tuple([(max(0, -p[0]), max(0, p[1]-self.shape[i])) for i,p in enumerate(arg_)])
-    return self.pad(padding).shrink(tuple([(p[0] + padding[i][0], p[1] + padding[i][0]) for i,p in enumerate(arg_)]))
+  def slice(self, arg: Sequence[Optional[Tuple[int, int]]], value:float=0) -> Tensor:
```
Comment (geohot) [inline tinygrad/tensor.py:254]: Why are there whitespace changes here?

Comment (ZenginU) [reply tinygrad/tensor.py]: Sorry formatted paste, fixed

**Comment by geohot** [comment]
Thanks!

**Pull request merged**